### PR TITLE
Added tool for custom user experts

### LINKED ISSUE #178: Better support for custom experts in Server
As of right now, hivemind.Server supports arbitrary expert class with fixed input sizes, but it is impossible to deploy custom experts with `hivemind-server` script. As a result, users who need custom experts must launch hivemind.Server manually from python.

We must refactor the server creator to support custom experts:
* [ ] Implement `register_expert` decorator in `hivemind/layers/register.py` that adds a given expert
```(python)
@hivemind.register_exprert('my_expert_name', sample_inputs=lambda hid_dim: (torch.randn(3, 128, hid_dim))
class MyCustomExpert(torch.nn.Module)
   ...
```
* [ ] make sure that the decorator returns an unmodified expert module
* [ ] add an assertion that forbids registering multiple experts with the same name
* [ ] re-write `hivemind/server/layers` to use `register_expert` instead of manually adding experts to global dictionaries
* [ ] implement `get_expert` method in `hivemind/layers/register.py`
* [ ] in `hivemind-server` script, add an option to import a custom module (that may or may not register experts inside)
* [ ] write a unit test in `hivemind/tests/test_moe.py` that creates a server with a custom expert, then runs forward and backward for this expert

### PR DESCRIPTION
**Pull request by romakail**
- This commit fixes #178 
- The user needs to wrap his own nn.module into a special decorator
- Added new command-line argument for run_server.py which indicates the path to file with user modules

### CODE REVIEW

File: hivemind/server/__init__.py [hunk 1]
```diff
@@ -103,7 +103,7 @@ def create(listen_on='0.0.0.0:*', num_experts: int = None, expert_uids: str = No
         """
         if len(kwargs) != 0:
             logger.info("Ignored kwargs:", kwargs)
-        assert expert_cls in name_to_block
+        # assert expert_cls in name_to_block
```
Comment (justheuristic) [inline hivemind/server/__init__.py:106]: please either fix or un-comment this

**Review by justheuristic — APPROVED** [review]
Requests before merge:
* add a simple test that creates a custom expert (registered right inside that test)
* un-comment that assert statement

File: hivemind/hivemind_cli/run_server.py [hunk 2]
```diff
@@ -11,6 +13,11 @@
 
 logger = get_logger(__name__)
 
+def add_custom_models_from_file(path):
+    spec = importlib.util.spec_from_file_location(
+        "custm_module", os.path.abspath(path))
```
Comment (mryab) [inline hivemind/hivemind_cli/run_server.py:18]:
```suggestion
        "custom_module", os.path.abspath(path))
```

File: hivemind/hivemind_cli/run_server.py [hunk 3]
```diff
@@ -47,6 +54,9 @@ def main():
     parser.add_argument('--checkpoint_dir', type=Path, required=False, help='Directory to store expert checkpoints')
     parser.add_argument('--load_experts', action='store_true', help='Load experts from the checkpoint directory')
 
+    parser.add_argument('--custom_module_path', type=str, default=None, required=False,
+                        help='Path of a file with cutom nn.modules, wrapped into special decorator')
```
Comment (mryab) [inline hivemind/hivemind_cli/run_server.py:58]:
```suggestion
                        help='Path of a file with custom nn.modules, wrapped into special decorator')
```

File: hivemind/server/module_registrator.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import typing as tp
+import functools
+import torch
+import torch.nn as nn
+
+from hivemind.server.layers import name_to_block, name_to_input
```
Comment (mryab) [inline hivemind/server/module_registrator.py:6]: Why not 
1) add this directly to hivemind.server.layers
2) register existing expert layers with this interface for uniformity?

File: hivemind/server/module_registrator.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    Adds a custom user expert to hivemind server.
+    :param name: the name of the expert. It shouldn't coincide with existing modules\
+        ('ffn', 'transforme', 'nop', 'det_dropout')
+    :param sample_input: a function which gets batch_size and hid_dim and outputs a \
+        sample of an input in the module
+    : unchanged module
+    """
+    def registrator(custom_class: tp.Type[nn.Module]):
+        if name in name_to_block or name in name_to_input:
+            raise RuntimeWarning("The class might already exist or be added twice")
```
Comment (mryab) [inline hivemind/server/module_registrator.py:19]: 1) We should probably raise an error in case of name collision
2) It can't be added twice, because the keys are unique

File: hivemind/server/module_registrator.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+def register_expert(name: str, sample_input: tp.Callable[[int, int], torch.tensor]):
+    """
+    Adds a custom user expert to hivemind server.
+    :param name: the name of the expert. It shouldn't coincide with existing modules\
+        ('ffn', 'transforme', 'nop', 'det_dropout')
+    :param sample_input: a function which gets batch_size and hid_dim and outputs a \
+        sample of an input in the module
+    : unchanged module
+    """
+    def registrator(custom_class: tp.Type[nn.Module]):
```
Comment (mryab) [inline hivemind/server/module_registrator.py:17]: It's a function, not a class/entity, so it's probably best to rename this to something like `register_expert_class`

**Review by mryab — CHANGES_REQUESTED** [review]

File: hivemind/hivemind_cli/run_server.py (hunk 2 above)
Comment (romakail) [reply hivemind/hivemind_cli/run_server.py]: Fixed

File: hivemind/server/layers/__init__.py [hunk 7]
```diff
@@ -1,19 +1,15 @@
...
+
 from hivemind.server.layers.common import FeedforwardBlock, TransformerEncoderLayer, NopExpert
 from hivemind.server.layers.dropout import DeterministicDropout, DeterministicDropoutNetwork
 from hivemind.server.layers.lr_schedule import get_linear_schedule_with_warmup
 
-name_to_block = {'ffn': lambda hid_dim: FeedforwardBlock(hid_dim),
-                 'transformer': lambda hid_dim: TransformerEncoderLayer(hid_dim, dim_feedforward=4 * hid_dim, nhead=16),
-                 'nop': lambda hid_dim: NopExpert(hid_dim),
-                 'det_dropout': lambda hid_dim: DeterministicDropoutNetwork(hid_dim, dropout_prob=0.2)}
+from hivemind.server.layers.custom_experts import *
```
Comment (mryab) [inline hivemind/server/layers/__init__.py:10]: No wildcard imports should be used, if possible

File: hivemind/server/layers/common.py [hunk 8]
```diff
@@ -1,13 +1,15 @@
...
 
+from hivemind.server.layers.custom_experts import register_expert_class
 
 # https://github.com/huggingface/transformers/blob/master/src/transformers/activations.py
 @torch.jit.script
 def gelu_fast(x):
     return 0.5 * x * (1.0 + torch.tanh(x * 0.7978845608 * (1.0 + 0.044715 * x * x)))
 
-
+ffn_sample_input = lambda batch_size, hid_dim: torch.empty((batch_size, hid_dim))
```
Comment (mryab) [inline hivemind/server/layers/common.py:11]: Extra lines are missing

File: tests/test_custom_expert.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+            loss = output0.sum()
+            loss.backward()
+            loss = output1.sum()
+            loss.backward()
+
+if __name__ == "__main__":
+    test_custom_expert()
+    with open('debug_logs', 'a') as f:
+        print("Done first : ", name, os.getpid(), file=f)
+    test_multihead_expert()
```
Comment (mryab) [inline tests/test_custom_expert.py:55]: Is this necessary?

File: tests/test_custom_expert.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        for step in range(1, 4):
+            batch = (torch.randn(step, hid_dim), torch.randn(step, 2 * hid_dim), torch.randn(step, 3 * hid_dim))
+
+            output0 = expert0(*batch)
+            output1 = expert1(*batch)
+
+            loss = output0.sum()
+            loss.backward()
+            loss = output1.sum()
+            loss.backward()
```
Comment (mryab) [inline tests/test_custom_expert.py:49]: Is this test necessary in presence of `test_custom_expert` or vice versa?

File: tests/test_custom_expert.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    with background_server(
+        expert_cls='perceptron', num_experts=2, device='cpu',
+        hidden_dim=hid_dim, num_handlers=2, no_dht=True,
+        custom_module_path=os.path.join(os.path.dirname(__file__), 'custom_networks.py')) as (server_endpoint, _):
+
+        expert0 = RemoteExpert('expert.0', server_endpoint)
+        expert1 = RemoteExpert('expert.1', server_endpoint)
+
+        for step in range(1, 4):
+            batch = torch.randn(step, hid_dim)
```
Comment (mryab) [inline tests/test_custom_expert.py:20]: 1) Why is batch size step-dependent?
2) Why do steps go from 1 to 4?
3) Isn't just one step sufficient?

File: hivemind/server/layers/custom_experts.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+def register_expert_class(name: str, sample_input: Callable[[int, int], torch.tensor]):
+    """
+    Adds a custom user expert to hivemind server.
+    :param name: the name of the expert. It shouldn't coincide with existing modules\
+        ('ffn', 'transformer', 'nop', 'det_dropout')
+    :param sample_input: a function which gets batch_size and hid_dim and outputs a \
+        sample of an input in the module
+    :unchanged module
+    """
+    def registrator(custom_class: Type[nn.Module]):
```
Comment (mryab) [inline hivemind/server/layers/custom_experts.py:24]: It's a function, not a class/entity, so it's probably best to rename this

File: hivemind/server/layers/custom_experts.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
+import os
+import importlib
+import torch
+import torch.nn as nn
+from typing import Callable, Type
+
+from hivemind.server.layers import name_to_block, name_to_input
```
Comment (mryab) [inline hivemind/server/layers/custom_experts.py:7]: Imports aren't sorted

File: hivemind/hivemind_cli/run_server.py [hunk 14]
```diff
@@ -1,8 +1,10 @@
 from functools import partial
 from pathlib import Path
 
+import os
 import configargparse
 import torch
+import importlib
```
Comment (mryab) [inline hivemind/hivemind_cli/run_server.py:8]: Imports aren't required

File: tests/test_custom_expert.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
+import os
+import pytest
+import torch
+import hivemind
+
+from typing import Optional
+from hivemind import RemoteExpert, background_server
```
Comment (mryab) [inline tests/test_custom_expert.py:7]: Imports aren't sorted

**Review by mryab — CHANGES_REQUESTED** [review]

File: hivemind/server/layers/custom_experts.py (hunk 12 above)
Comment (justheuristic) [reply hivemind/server/layers/custom_experts.py:24]: e.g. register_expert

File: hivemind/server/layers/custom_experts.py (hunk 13 above)
Comment (justheuristic) [reply hivemind/server/layers/custom_experts.py:7]: https://donjayamanne.github.io/pythonVSCodeDocs/docs/refactoring_sort-imports/

File: tests/test_custom_expert.py (hunk 11 above)
Comment (justheuristic) [reply tests/test_custom_expert.py:20]: Interpretation: it may be sufficient to check just two batch sizes: 1 and not_1

File: hivemind/server/layers/__init__.py (hunk 7 above)
Comment (justheuristic) [reply hivemind/server/layers/__init__.py:10]: ref: https://www.python.org/dev/peps/pep-0008/#id23

File: tests/test_custom_expert.py (hunk 10 above)
Comment (justheuristic) [reply tests/test_custom_expert.py:49]: If i remember the counterargument correctly:
- this test verifies that custom experts support multiple inputs
- this test verifies that custom expert inputs handle higher ndims
- this test verifies that the new code can handle multiple custom experts correctly

File: hivemind/server/layers/common.py (hunk 8 above)
Comment (justheuristic) [reply hivemind/server/layers/common.py:11]: reference: https://www.python.org/dev/peps/pep-0008/#blank-lines

File: tests/test_custom_expert.py (hunk 10 above)
Comment (romakail) [reply tests/test_custom_expert.py:49]: I was checking that custom expert works correctly with inputs consisting of multiple tensors e.g. transformer's layers which take input and mask

File: tests/test_custom_expert.py (hunk 15 above)
Comment (romakail) [reply tests/test_custom_expert.py]: Fixed

File: hivemind/hivemind_cli/run_server.py (hunk 14 above)
Comment (romakail) [reply hivemind/hivemind_cli/run_server.py]: Deleted

File: hivemind/server/layers/custom_experts.py (hunk 13 above)
Comment (romakail) [reply hivemind/server/layers/custom_experts.py]: Fixed

File: hivemind/server/layers/custom_experts.py (hunk 12 above)
Comment (romakail) [reply hivemind/server/layers/custom_experts.py]: renamed

File: tests/test_custom_expert.py (hunk 11 above)
Comment (romakail) [reply tests/test_custom_expert.py]: Fixed

File: tests/test_custom_expert.py (hunk 9 above)
Comment (romakail) [reply tests/test_custom_expert.py]: Deleted. Didn't remove after debug/

File: hivemind/server/layers/common.py (hunk 8 above)
Comment (romakail) [reply hivemind/server/layers/common.py:13]: Added lines in both files

File: hivemind/server/layers/__init__.py (hunk 7 above)
Comment (romakail) [reply hivemind/server/layers/__init__.py]: Fixed

File: hivemind/hivemind_cli/run_server.py [hunk 16]
```diff
@@ -56,6 +56,9 @@ def main():
     parser.add_argument('--stats_report_interval', type=int, required=False,
                         help='Interval between two reports of batch processing performance statistics')
 
+    parser.add_argument('--custom_module_path', type=str, default=None, required=False,
```
Comment (mryab) [inline hivemind/hivemind_cli/run_server.py:59]: Specifying `default` isn't necessary when required=False

File: hivemind/server/layers/__init__.py [hunk 17]
```diff
@@ -1,19 +1,15 @@
 import torch
 
+name_to_block = {}
+name_to_input = {}
+
 from hivemind.server.layers.common import FeedforwardBlock, TransformerEncoderLayer, NopExpert
 from hivemind.server.layers.dropout import DeterministicDropout, DeterministicDropoutNetwork
```
Comment (mryab) [inline hivemind/server/layers/__init__.py:7]: These imports are now unnecessary

File: hivemind/server/layers/common.py [hunk 18]
```diff
@@ -58,7 +63,20 @@ def forward(self, src, src_key_padding_mask=None):
...
 
 
+transformer_sample_input = lambda batch_size, hid_dim: \
+    (torch.empty((batch_size, 128, hid_dim)), \
+    torch.empty((batch_size, 128), dtype=torch.bool))
+@register_expert_class('transformer', transformer_sample_input)
+class TunedTransformer(TransformerEncoderLayer):
+
+    def __init__(self, hid_dim):
+        super(TunedTransformer, self).__init__(hid_dim, dim_feedforward=4 * hid_dim, nhead=16)
```
Comment (mryab) [inline hivemind/server/layers/common.py:73]:
```suggestion
        super().__init__(hid_dim, dim_feedforward=4 * hid_dim, nhead=16)
```

File: tests/custom_networks.py [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+        super(MultilayerPerceptron, self).__init__()
+        self.layer1 = nn.Linear(hidden_dim, 2 * hidden_dim)
+        self.layer2 = nn.Linear(2 * hidden_dim, 2 * hidden_dim)
+        self.layer3 = nn.Linear(2 * hidden_dim, num_classes)
+
+    def forward(self, x):
+        x = F.relu(self.layer1(x))
+        x = F.relu(self.layer2(x))
+        x = self.layer3(x)
+        return x
```
Comment (mryab) [inline tests/custom_networks.py:20]: I agree about multi-input and loading from separate files, but then the perceptron test isn't necessary, is it?

File: tests/test_custom_expert.py [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+def test_multihead_expert(port: Optional[int] = None, hid_dim=16):
+    with background_server(
+        expert_cls='multihead', num_experts=2, device='cpu',
+        hidden_dim=hid_dim, num_handlers=2, no_dht=True,
+        custom_module_path=os.path.join(os.path.dirname(__file__), 'custom_networks.py')) as (server_endpoint, _):
+
+        expert0 = RemoteExpert('expert.0', server_endpoint)
+        expert1 = RemoteExpert('expert.1', server_endpoint)
+
+        for batch_size in (1, 4):
```
Comment (mryab) [inline tests/test_custom_expert.py:41]: Not sure if it's fixed, still see 3 batch sizes. For checking custom experts even 1 (maybe 2 for the worst case) are sufficient

**Review by mryab — CHANGES_REQUESTED** [review]

File: tests/custom_networks.py (hunk 19 above)
Comment (mryab) [reply tests/custom_networks.py:20]: Nevermind, I misunderstood the last part about several custom experts. Thank you!

File: hivemind/hivemind_cli/run_server.py (hunk 16 above)
Comment (romakail) [reply hivemind/hivemind_cli/run_server.py:59]: Fixed. Yet there are four more parser parameters which have default=None and required=False specified.

File: hivemind/server/layers/__init__.py (hunk 17 above)
Comment (romakail) [reply hivemind/server/layers/__init__.py:7]: Deleted

File: hivemind/server/layers/common.py (hunk 18 above)
Comment (romakail) [reply hivemind/server/layers/common.py:73]: Fixed

File: tests/test_custom_expert.py (hunk 20 above)
Comment (romakail) [reply tests/test_custom_expert.py:41]: It's a tuple, not a range, which means I check only batch_size equal to 1 and 4.
To sum up, what I attempt to test:
- custom_experts work correctly bot with batch_size equal to 1 and not 1
- different custom_experts do not conflict if they are defined in one file
- custom experts fine well with any kind of input e.g. multiple tensors

Comment (romakail) [reply tests/test_custom_expert.py:41]: My two tests cover all of these aspects, but I'm not sure they do so in the most efficient, brief, and fancy way.

**Review by mryab — APPROVED** [review]
